servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherProcessor.java

```diff
@@ -0,0 +1,259 @@
+        }
+    }
+
+    private void tryEmitSignals() {
+        boolean tryAcquire = true;
+        while (tryAcquire && tryAcquireLock(emittingUpdater, this)) {
+            final BufferConsumer<T> consumer = this.consumer;
+            try {
+                if (consumer instanceof SubscriberBufferConsumer) {
+                    SubscriberBufferConsumer<T> target = (SubscriberBufferConsumer<T>) consumer;
+                    if (fatalError != null) {
+                        earlyTerminateConsumerHoldingLock(target, fatalError);
+                        return;
+                    } else {
+                        emitSignalsHoldingLock(target);
+                    }
+                }
+            } finally {
+                tryAcquire = !releaseLock(emittingUpdater, this);
+            }
+        }
+    }
+
+    private void emitSignalsHoldingLock(final SubscriberBufferConsumer<T> target) {
+        for (;;) {
+            final long cPending = pending;
+            if (cPending > 0 && pendingUpdater.compareAndSet(this, cPending, cPending - 1)) {
+                final boolean consumed;
+                try {
+                    consumed = buffer.tryConsume(target);
+                } catch (Throwable t) {
+                    earlyTerminateConsumerHoldingLock(target, t);
+                    return;
+                }
+
+                if (target.isTerminated()) {
+                    pending = Long.MIN_VALUE;
+                } else if (!consumed) {
+                    // we optimistically decremented pending, so increment back again.
+                    pendingUpdater.accumulateAndGet(this, 1,
+                            FlowControlUtils::addWithOverflowProtectionIfNotNegative);
+                    return;
+                }
+            } else if (cPending < 0) {
+                // cancelled or already terminated
+                return;
+            } else if (cPending == 0) {
+                final boolean consumed;
+                try {
+                    consumed = buffer.tryConsumeTerminal(target);
+                } catch (Throwable t) {
+                    // Assume that we did not deliver terminal to the consumer.
+                    earlyTerminateConsumerHoldingLock(target, t);
+                    return;
+                }
+                if (consumed) {
+                    pending = Long.MIN_VALUE;
+                    return;
+                }
+                break;
+            }
+        }
+    }
+
+    private void earlyTerminateConsumerHoldingLock(final SubscriberBufferConsumer<T> consumer, final Throwable cause) {
+        pending = Long.MIN_VALUE;
+        try {
+            delayedSubscription.cancel();
+        } finally {
+            consumer.consumeTerminal(cause);
+        }
+    }
+
+    private static final class SubscriberBufferConsumer<T> implements BufferConsumer<T> {
+        private final Subscriber<? super T> subscriber;
+        private boolean terminated;
```

Comment:
does this need to be volatile ?

Reply:
nope, this class ins't used concurrently by different threads.